WordPress/Classes/Utility/Media/MediaAssetExporter.swift

```diff
@@ -0,0 +1,236 @@
+        } else {
+            onError(AssetExportError.unsupportedPHAssetMediaType)
+        }
+    }
+
+    fileprivate func exportImage(forAsset asset: PHAsset, onCompletion: @escaping (AssetExport) -> Void, onError: @escaping (MediaExportError) -> Void) {
+        do {
+            guard asset.mediaType == .image else {
+                throw AssetExportError.expectedPHAssetImageType
+            }
+
+            // Get the resource matching the type, to export.
+            let resources = PHAssetResource.assetResources(for: asset).filter({ $0.type == .photo })
+            guard let resource = resources.first else {
+                throw AssetExportError.unavailablePHAssetImageResource
+            }
+
+            if UTTypeEqual(resource.uniformTypeIdentifier as CFString, kUTTypeGIF) {
+                // Since this is a GIF, handle the export in it's own way.
+                exportGIF(forAsset: asset, resource: resource, onCompletion: onCompletion, onError: onError)
+                return
+            }
+
+            // Configure the options for requesting the image.
+            let options = PHImageRequestOptions()
+            options.version = .current
+            options.deliveryMode = .highQualityFormat
+            options.resizeMode = .exact
+            options.isNetworkAccessAllowed = true
+
+            // Configure the targetSize for PHImageManager to resize to.
+            let targetSize: CGSize
+            if let maximumImageSize = maximumImageSize {
+                targetSize = CGSize(width: maximumImageSize, height: maximumImageSize)
+            } else {
+                targetSize = PHImageManagerMaximumSize
+            }
+
+            // Configure an error handler for the image request.
+            let onImageRequestError: (Error?) -> Void = { (error) in
+                guard let error = error else {
+                    onError(AssetExportError.failedLoadingPHImageManagerRequest)
+                    return
+                }
+                onError(self.exporterErrorWith(error: error))
+            }
+
+            // Request the image.
+            let manager = PHImageManager.default()
+            manager.requestImage(for: asset,
+                                 targetSize: targetSize,
+                                 contentMode: .aspectFit,
+                                 options: options,
+                                 resultHandler: { (image, info) in
+                                    guard let image = image else {
+                                        onImageRequestError(info?[PHImageErrorKey] as? Error)
+                                        return
+                                    }
+                                    // Hand off the image export to a shared image writer.
+                                    let exporter = MediaImageExporter()
+                                    exporter.maximumImageSize = self.maximumImageSize
```

Comment:
I've looked at this for a good bit here, and tested with the new `MediaLibrary` code. At the moment, due to the block ownership and un-ownership via the shared `PHImageManager` we do not create retain cycles.

However, I'm not 100% convinced it's safe, so I'd like to keep an eye on it with the new MediaLibrary code and test for leaks with the complete integration.